File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/I-Что_Да_Пузырек/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    seen: list[bool] = [False] * (n + 1)
    last: int = n
    zeros: int = 0
    sys.stdout.write('1')
    for num in nums:
        sys.stdout.write(' ')
        seen[num] = True
        zeros += 1
        while seen[last]:
            last -= 1
            zeros -= 1
        sys.stdout.write(str(zeros + 1))
    sys.stdout.write('\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/I-Что_Да_Пузырек/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<bool> seen (n + 1, false);
    std::vector<int> nums (n, 0);
    for (int i = 0; i < n; ++i) std::cin >> nums[i];
    int zeros = 0;
    int last = n;
    std::cout << '1';
    for (int& num : nums) {
        std::cout << " ";
        seen[num] = true;
        ++zeros;
        while (seen[last]) {
            --last;
            --zeros;
        }
        std::cout << zeros + 1;
    }
    std::cout << "\n";
}


int main() {
    solution();
}